### PR TITLE
[5.0] Add Option to use L3 HA with Keepalived for Neutron router failover

### PR DESCRIPTION
**Pull request by dirkmueller**
Backport-Of: #2005

### CODE REVIEW

File: chef/cookbooks/ceilometer/recipes/central_ha.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@
 end
 
 crowbar_pacemaker_order_only_existing "o-#{service_name}" do
-  ordering "( #{rabbit_settings[:pacemaker_resource]} cl-keystone ) #{service_name}"
+  ordering "( galera #{rabbit_settings[:pacemaker_resource]} ) #{service_name}"
```
Comment (sjamgade) [inline chef/cookbooks/ceilometer/recipes/central_ha.rb]: should'nt this be `ms-galera` ?

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 2]
```diff
@@ -246,68 +246,82 @@
...
+    rabbit_settings = fetch_rabbitmq_settings
+
+    crowbar_pacemaker_order_only_existing "o-#{ha_service_primitive_name}" do
+      # While neutron-ha-tool technically doesn't directly depend on galera or
+      # rabbitmq, if these bits are not running, then neutron-server can run but
+      # can't do what it's being asked. Note that neutron-server does have a
+      # constraint on these services, but it's optional, not mandatory (because it
+      # doesn't need to be restarted when galera or rabbitmq are restarted).
+      # So explicitly depend on galera and rabbitmq (if they are in the cluster).
+      ordering "( galera #{rabbit_settings[:pacemaker_resource]} g-haproxy cl-neutron-server " \
```
Comment (sjamgade) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: ditto

File: chef/cookbooks/nova/recipes/compute_ha.rb [hunk 3]
```diff
@@ -334,9 +334,7 @@
   #  - cinder is used in case of boot from volume
   #  - neutron agents are used even with DVR, if only to have a DHCP server for
   #    the instance to get an IP address
-  ordering "( postgresql #{rabbit_settings[:pacemaker_resource]} cl-keystone cl-swift-proxy " \
-      "cl-glance-api cl-cinder-api cl-neutron-server cl-neutron-dhcp-agent cl-neutron-l3-agent " \
-      "cl-neutron-metadata-agent cl-nova-api ) #{evacuate_primitive}"
+  ordering "( galera #{rabbit_settings[:pacemaker_resource]} ) #{evacuate_primitive}"
```
Comment (sjamgade) [inline chef/cookbooks/nova/recipes/compute_ha.rb]: ditto

Comment (dirkmueller) [reply chef/cookbooks/nova/recipes/compute_ha.rb]: fixed

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb (hunk 2 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: fixed

File: chef/cookbooks/ceilometer/recipes/central_ha.rb (hunk 1 above)
Comment (dirkmueller) [reply chef/cookbooks/ceilometer/recipes/central_ha.rb]: fixed

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 4]
```diff
@@ -246,68 +246,82 @@
...
+  ha_service_primitive_name = "neutron-l3-ha-service"
 
-  # Remove old location
-  ha_tool_location_name = "l-#{ha_tool_primitive_name}-controller"
-  pacemaker_location ha_tool_location_name do
-    action :delete
-    only_if "crm configure show #{ha_tool_location_name}"
-    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
-  end
+  if node[:neutron][:l3_ha][:use_l3_ha]
```
Comment (AbelNavarro) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: The logic of this should be reversed. If we have `:use_l3_ha` it makes no sense to delete it and do nothing more.

**Comment by AbelNavarro** [comment]
[L3 HA](https://trello.com/c/OJyN9gjj/155-l3-ha)

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb (hunk 4 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: what else needs to be done other than deleting teh neutron-ha-tool ressource?

Comment (AbelNavarro) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: sorry, got confused while checking the original code. LGTM.

**Pull request merged**